File: app/src/main/java/com/websarva/wings/android/zuboradiary/ui/settings/SettingsViewModel.java
package com.websarva.wings.android.zuboradiary.ui.settings;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.datastore.preferences.core.Preferences;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.websarva.wings.android.zuboradiary.data.AppMessage;
import com.websarva.wings.android.zuboradiary.data.database.DiaryRepository;
import com.websarva.wings.android.zuboradiary.data.network.GeoCoordinates;
import com.websarva.wings.android.zuboradiary.data.preferences.CalendarStartDayOfWeekPreference;
import com.websarva.wings.android.zuboradiary.data.preferences.UserPreferencesRepository;
import com.websarva.wings.android.zuboradiary.data.preferences.WeatherInfoAcquisitionPreference;
import com.websarva.wings.android.zuboradiary.data.preferences.PassCodeLockPreference;
import com.websarva.wings.android.zuboradiary.data.preferences.ReminderNotificationPreference;
import com.websarva.wings.android.zuboradiary.data.preferences.ThemeColorPreference;
import com.websarva.wings.android.zuboradiary.data.preferences.ThemeColor;
import com.websarva.wings.android.zuboradiary.data.worker.WorkerRepository;
import com.websarva.wings.android.zuboradiary.ui.BaseViewModel;

import java.time.DayOfWeek;
import java.time.LocalTime;
import java.util.Objects;
import java.util.concurrent.CancellationException;
import java.util.concurrent.ExecutionException;

import javax.inject.Inject;

import dagger.hilt.android.lifecycle.HiltViewModel;
import io.reactivex.rxjava3.core.Flowable;
import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.disposables.CompositeDisposable;
import io.reactivex.rxjava3.functions.Consumer;

@HiltViewModel
public class SettingsViewModel extends BaseViewModel {
    private final UserPreferencesRepository userPreferencesRepository;
    private final WorkerRepository workerRepository;
    private final DiaryRepository diaryRepository;
    private final CompositeDisposable disposables = new CompositeDisposable();

    private final MutableLiveData<ThemeColor> themeColor = new MutableLiveData<>();
    private final MutableLiveData<DayOfWeek> calendarStartDayOfWeek = new MutableLiveData<>();
    private final MutableLiveData<Boolean> isCheckedReminderNotification = new MutableLiveData<>();
    private final MutableLiveData<LocalTime> reminderNotificationTime = new MutableLiveData<>();
    private final MutableLiveData<Boolean> isCheckedPasscodeLock = new MutableLiveData<>();
    private final MutableLiveData<Boolean> isCheckedWeatherInfoAcquisition = new MutableLiveData<>();

    private final MutableLiveData<GeoCoordinates> geoCoordinates = new MutableLiveData<>();
    private Flowable<ThemeColorPreference> themeColorPreferenceFlowable;
    private Flowable<CalendarStartDayOfWeekPreference> calendarStartDayPreferenceFlowable;
    private Flowable<ReminderNotificationPreference> reminderNotificationPreferenceFlowable;
    private Flowable<PassCodeLockPreference> passCodeLockPreferenceFlowable;
    private Flowable<WeatherInfoAcquisitionPreference> weatherInfoAcquisitionPreferenceFlowable;

    @Inject
    public SettingsViewModel(
            UserPreferencesRepository userPreferencesRepository,
            WorkerRepository workerRepository,
            DiaryRepository diaryRepository) {
        this.userPreferencesRepository = userPreferencesRepository;
        this.workerRepository = workerRepository;
        this.diaryRepository = diaryRepository;

        initialize();
    }

    @Override
    protected void initialize() {
        initializeAppMessageList();
        setUpThemeColorPreferenceValueLoading();
        setUpCalendarStartDayOfWeekPreferenceValueLoading();
        setUpReminderNotificationPreferenceValueLoading();
        setUpPasscodeLockPreferenceValueLoading();
        setUpWeatherInfoAcquisitionPreferenceValueLoading();
    }

    private void setUpThemeColorPreferenceValueLoading() {
        themeColorPreferenceFlowable = userPreferencesRepository.loadThemeColorPreference();
        disposables.add(
                themeColorPreferenceFlowable.subscribe(
                        value -> {
                            // HACK:一つのDataStore(UserPreferencesクラス)からFlowableを生成している為、
                            //      一つのPreferenceを更新すると他のPreferenceのFlowableにも通知される。
                            //      結果的にObserverにも通知が行き、不必要な処理が発生してしまう。
                            //      対策として下記コードを記述。(他PreferenceFlowableも同様)
                            Objects.requireNonNull(value);
                            ThemeColor themeColor = this.themeColor.getValue();
                            if (themeColor != null && themeColor.equals(value.toThemeColor())) return;

                            this.themeColor.postValue(value.toThemeColor());
                        },
                        throwable -> {
                            Log.d("Exception", "テーマカラー設定値読込失敗", throwable);
                            addSettingLoadingErrorMessage();
                        }
                )
        );
    }

    @NonNull
    public ThemeColor loadThemeColorSettingValue() {
        ThemeColor themeColorValue = themeColor.getValue();
        if (themeColorValue != null) return themeColorValue;
        ThemeColorPreference defaultValue = new ThemeColorPreference();
        return themeColorPreferenceFlowable.blockingFirst(defaultValue).toThemeColor();
    }

    private void setUpCalendarStartDayOfWeekPreferenceValueLoading() {
        calendarStartDayPreferenceFlowable =
                userPreferencesRepository.loadCalendarStartDayOfWeekPreference();
        disposables.add(
                calendarStartDayPreferenceFlowable.subscribe(
                        value -> {
                            Objects.requireNonNull(value);
                            DayOfWeek dayOfWeek = calendarStartDayOfWeek.getValue();
                            if (dayOfWeek != null && dayOfWeek.equals(value.toDayOfWeek())) return;

                            DayOfWeek calendarStartDayOfWeek = value.toDayOfWeek();
                            this.calendarStartDayOfWeek.postValue(calendarStartDayOfWeek);
                            },
                        throwable -> {
                            Log.d("Exception", "カレンダー開始曜日設定値読込失敗", throwable);
                            addSettingLoadingErrorMessage();
                        }
                )
        );
    }

    @NonNull
    public DayOfWeek loadCalendarStartDaySettingValue() {
        DayOfWeek dayOfWeekValue = calendarStartDayOfWeek.getValue();
        if (dayOfWeekValue != null) return dayOfWeekValue;
        CalendarStartDayOfWeekPreference defaultValue = new CalendarStartDayOfWeekPreference();
        return calendarStartDayPreferenceFlowable.blockingFirst(defaultValue).toDayOfWeek();
    }

    private void setUpReminderNotificationPreferenceValueLoading() {
        reminderNotificationPreferenceFlowable =
                userPreferencesRepository.loadReminderNotificationPreference();
        disposables.add(
                reminderNotificationPreferenceFlowable.subscribe(
                        value -> {
                            Objects.requireNonNull(value);
                            Boolean isChecked = isCheckedReminderNotification.getValue();
                            if (isChecked != null && isChecked == value.getIsChecked()) return;

                            isCheckedReminderNotification.postValue(value.getIsChecked());
                            reminderNotificationTime.postValue(value.getNotificationLocalTime());
                            },
                        throwable -> {
                            Log.d("Exception", "リマインダー通知設定値読込失敗", throwable);
                            addSettingLoadingErrorMessage();
                        }
                )
        );
    }

    public boolean isCheckedReminderNotificationSetting() {
        Boolean value = isCheckedReminderNotification.getValue();
        if (value != null) return value;
        ReminderNotificationPreference defaultValue =
                new ReminderNotificationPreference();
        return reminderNotificationPreferenceFlowable.blockingFirst(defaultValue).getIsChecked();
    }

    @Nullable
    public LocalTime loadReminderNotificationTimeSettingValue() {
        LocalTime value = reminderNotificationTime.getValue();
        if (value != null) return value;
        ReminderNotificationPreference defaultValue =
                new ReminderNotificationPreference();
        return reminderNotificationPreferenceFlowable
                .blockingFirst(defaultValue).getNotificationLocalTime();
    }

    private void setUpPasscodeLockPreferenceValueLoading() {
        passCodeLockPreferenceFlowable = userPreferencesRepository.loadPasscodeLockPreference();
        disposables.add(
                passCodeLockPreferenceFlowable.subscribe(
                        value -> {
                            Objects.requireNonNull(value);
                            Boolean isChecked = isCheckedPasscodeLock.getValue();
                            if (isChecked != null && isChecked == value.getIsChecked()) return;

                            isCheckedPasscodeLock.postValue(value.getIsChecked());
                        },
                        throwable -> {
                            Log.d("Exception", "パスコード設定値読込失敗", throwable);
                            addSettingLoadingErrorMessage();
                        }
                )
        );
    }

    private void setUpWeatherInfoAcquisitionPreferenceValueLoading() {
        weatherInfoAcquisitionPreferenceFlowable =
                userPreferencesRepository.loadWeatherInfoAcquisitionPreference();
        disposables.add(
                weatherInfoAcquisitionPreferenceFlowable.subscribe(
                        value -> {
                            Objects.requireNonNull(value);
                            Boolean isChecked = isCheckedWeatherInfoAcquisition.getValue();
                            if (isChecked != null && isChecked == value.getIsChecked()) return;

                            isCheckedWeatherInfoAcquisition.postValue(value.getIsChecked());
                        },
                        throwable -> {
                            Log.d("Exception", "天気情報取得設定値読込失敗", throwable);
                            addSettingLoadingErrorMessage();
                        }
                )
        );
    }

    public boolean isCheckedWeatherInfoAcquisitionSetting() {
        Boolean value = isCheckedWeatherInfoAcquisition.getValue();
        if (value != null) return value;
        WeatherInfoAcquisitionPreference defaultValue =
                new WeatherInfoAcquisitionPreference();
        return weatherInfoAcquisitionPreferenceFlowable.blockingFirst(defaultValue).getIsChecked();
    }

    private void addSettingLoadingErrorMessage() {
        if (equalLastAppMessage(AppMessage.SETTING_LOADING_ERROR)) return;  // 設定更新エラー通知の重複防止
        addAppMessage(AppMessage.SETTING_LOADING_ERROR);
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        disposables.clear();
    }

    void saveThemeColor(ThemeColor value) {
        Objects.requireNonNull(value);

        ThemeColorPreference preferenceValue = new ThemeColorPreference(value);
        Single<Preferences> result = userPreferencesRepository.saveThemeColorPreference(preferenceValue);
        setUpProcessOnUpdate(result, null);
    }

    void saveCalendarStartDayOfWeek(DayOfWeek value) {
        Objects.requireNonNull(value);

        CalendarStartDayOfWeekPreference preferenceValue =
                new CalendarStartDayOfWeekPreference(value);
        Single<Preferences> result =
                userPreferencesRepository.saveCalendarStartDayOfWeekPreference(preferenceValue);
        setUpProcessOnUpdate(result, null);
    }

    void saveReminderNotificationValid(LocalTime value) {
        Objects.requireNonNull(value);

        ReminderNotificationPreference preferenceValue =
                new ReminderNotificationPreference(true, value);
        Single<Preferences> result = userPreferencesRepository.saveReminderNotificationPreference(preferenceValue);
        setUpProcessOnUpdate(result, new OnSettingsUpdateCallback() {
            @Override
            public void onUpdateSettings() {
                workerRepository.registerReminderNotificationWorker(value);
            }
        });
    }

    public void saveReminderNotificationInvalid() {
        ReminderNotificationPreference preferenceValue =
                new ReminderNotificationPreference(false,(LocalTime) null);
        Single<Preferences> result = userPreferencesRepository.saveReminderNotificationPreference(preferenceValue);
        setUpProcessOnUpdate(result, new OnSettingsUpdateCallback() {
            @Override
            public void onUpdateSettings() {
                workerRepository.cancelReminderNotificationWorker();
            }
        });
    }

    void savePasscodeLock(boolean value) {
        String passcode;
        if (value) {
            passcode = "0000"; // TODO:仮
        } else {
            passcode = "";
        }

        PassCodeLockPreference preferenceValue = new PassCodeLockPreference(value, passcode);
        Single<Preferences> result = userPreferencesRepository.savePasscodeLockPreference(preferenceValue);
        setUpProcessOnUpdate(result, null);
    }

    public void saveWeatherInfoAcquisition(boolean value) {
        WeatherInfoAcquisitionPreference preferenceValue =
                new WeatherInfoAcquisitionPreference(value);
        Single<Preferences> result =
                userPreferencesRepository.saveWeatherInfoAcquisitionPreference(preferenceValue);
        setUpProcessOnUpdate(result, null);
    }

    @FunctionalInterface
    private interface OnSettingsUpdateCallback {
        void onUpdateSettings();
    }

    private void setUpProcessOnUpdate(Single<Preferences> result, @Nullable OnSettingsUpdateCallback callback) {
        Objects.requireNonNull(result);

        disposables.add(result.subscribe(new Consumer<Preferences>() {
            @Override
            public void accept(Preferences preferences) {
                Objects.requireNonNull(preferences);

                if (callback == null) return;
                callback.onUpdateSettings();
            }
        }, new Consumer<Throwable>() {
            @Override
            public void accept(Throwable throwable) {
                Objects.requireNonNull(throwable);

                AppMessage appMessage = AppMessage.SETTING_UPDATE_ERROR;
                if (equalLastAppMessage(appMessage)) return; // 設定更新エラー通知の重複防止
                addAppMessage(appMessage);
            }
        }));
    }

    public void updateGeoCoordinates(GeoCoordinates geoCoordinates) {
        Objects.requireNonNull(geoCoordinates);

        this.geoCoordinates.setValue(geoCoordinates);
    }

    public void clearGeoCoordinates() {
        geoCoordinates.setValue(null);
    }

    public boolean hasUpdatedGeoCoordinates() {
        GeoCoordinates geoCoordinates = this.geoCoordinates.getValue();
        return geoCoordinates != null;
    }

    void deleteAllDiaries() {
        try {
            diaryRepository.deleteAllDiaries().get();
        } catch (CancellationException | ExecutionException | InterruptedException e) {
            addAppMessage(AppMessage.DIARY_DELETE_ERROR);
        }
    }

    void deleteAllSettings() {
        Single<Preferences> result = userPreferencesRepository.initializeAllPreferences();
        setUpProcessOnUpdate(result, null);
    }

    void deleteAllData() {
        try {
            diaryRepository.deleteAllData().get();
        } catch (CancellationException | ExecutionException | InterruptedException e) {
            addAppMessage(AppMessage.DIARY_DELETE_ERROR);
        }
        deleteAllSettings();
    }

    // Getter/Setter
    public LiveData<ThemeColor> getThemeColorSettingValueLiveData() {
        return themeColor;
    }

    public LiveData<DayOfWeek> getCalendarStartDayOfWeekLiveData() {
        return calendarStartDayOfWeek;
    }

    public LiveData<Boolean> getIsCheckedReminderNotificationLiveData() {
        return isCheckedReminderNotification;
    }

    public LiveData<LocalTime> getReminderNotificationTimeLiveData() {
        return reminderNotificationTime;
    }

    public LiveData<Boolean> getIsCheckedPasscodeLockLiveData() {
        return isCheckedPasscodeLock;
    }

    public LiveData<Boolean> getIsCheckedWeatherInfoAcquisitionLiveData() {
        return isCheckedWeatherInfoAcquisition;
    }

    public LiveData<GeoCoordinates> getGeoCoordinatesLiveData() {
        return geoCoordinates;
    }
}

File: app/src/main/java/com/websarva/wings/android/zuboradiary/ui/settings/SettingsViewModel.kt
package com.websarva.wings.android.zuboradiary.ui.settings

import android.util.Log
import androidx.datastore.core.IOException
import androidx.lifecycle.viewModelScope
import com.websarva.wings.android.zuboradiary.data.database.DiaryRepository
import com.websarva.wings.android.zuboradiary.data.network.GeoCoordinates
import com.websarva.wings.android.zuboradiary.data.preferences.CalendarStartDayOfWeekPreference
import com.websarva.wings.android.zuboradiary.data.preferences.PassCodeLockPreference
import com.websarva.wings.android.zuboradiary.data.preferences.ReminderNotificationPreference
import com.websarva.wings.android.zuboradiary.data.preferences.ThemeColor
import com.websarva.wings.android.zuboradiary.data.preferences.ThemeColorPreference
import com.websarva.wings.android.zuboradiary.data.preferences.UserPreferencesRepository
import com.websarva.wings.android.zuboradiary.data.preferences.WeatherInfoAcquisitionPreference
import com.websarva.wings.android.zuboradiary.data.worker.WorkerRepository
import com.websarva.wings.android.zuboradiary.createLogTag
import com.websarva.wings.android.zuboradiary.ui.appmessage.SettingsAppMessage
import com.websarva.wings.android.zuboradiary.ui.base.BaseViewModel
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import java.time.DayOfWeek
import java.time.LocalTime
import javax.inject.Inject

@HiltViewModel
class SettingsViewModel @Inject constructor(
    private val userPreferencesRepository: UserPreferencesRepository,
    private val workerRepository: WorkerRepository,
    private val diaryRepository: DiaryRepository
) : BaseViewModel() {

    private val logTag = createLogTag()

    // MEMO:StateFlow型設定値変数の値ははPreferencesDatastoreの値のみを代入したいので、
    //      代入されるまでの間(初回設定値読込中)はnullとする。
    lateinit var themeColor: StateFlow<ThemeColor?>
        private set
    lateinit var calendarStartDayOfWeek: StateFlow<DayOfWeek?>
        private set
    lateinit var isCheckedReminderNotification: StateFlow<Boolean?>
        private set
    lateinit var reminderNotificationTime: StateFlow<LocalTime?>
        private set
    lateinit var isCheckedPasscodeLock: StateFlow<Boolean?>
        private set
    private lateinit var passcode: StateFlow<String?>
    lateinit var isCheckedWeatherInfoAcquisition: StateFlow<Boolean?>
        private set

    private val _isAllSettingsNotNull = MutableStateFlow(false)
    val isAllSettingsNotNull
        get() = _isAllSettingsNotNull.asStateFlow()

    private val initialGeoCoordinates = null
    private val _geoCoordinates = MutableStateFlow<GeoCoordinates?>(initialGeoCoordinates)
    val geoCoordinates
        get() = _geoCoordinates.asStateFlow()

    val hasUpdatedGeoCoordinates
        get() = _geoCoordinates.value != null

    private val initialScrollPositionY = 0
    var scrollPositionY = initialScrollPositionY

    init {
        setUpThemeColorPreferenceValueLoading()
        setUpCalendarStartDayOfWeekPreferenceValueLoading()
        setUpReminderNotificationPreferenceValueLoading()
        setUpPasscodeLockPreferenceValueLoading()
        setUpWeatherInfoAcquisitionPreferenceValueLoading()
    }

    override fun initialize() {
        super.initialize()
        setUpThemeColorPreferenceValueLoading()
        setUpCalendarStartDayOfWeekPreferenceValueLoading()
        setUpReminderNotificationPreferenceValueLoading()
        setUpPasscodeLockPreferenceValueLoading()
        setUpWeatherInfoAcquisitionPreferenceValueLoading()
        clearGeoCoordinates()
        scrollPositionY = initialScrollPositionY
    }

    private fun setUpThemeColorPreferenceValueLoading() {
        loadSettingValue {
            themeColor =
                userPreferencesRepository.loadThemeColorPreference()
                    .map { preference ->
                        preference.themeColor
                    }.stateIn(
                        viewModelScope,
                        SharingStarted.Eagerly,
                        null
                    )
        }
        checkSettingNotNull(themeColor)
    }

    private fun setUpCalendarStartDayOfWeekPreferenceValueLoading() {
        loadSettingValue {
            calendarStartDayOfWeek =
                userPreferencesRepository.loadCalendarStartDayOfWeekPreference()
                    .map { preference ->
                        preference.dayOfWeek
                    }.stateIn(
                        viewModelScope,
                        SharingStarted.Eagerly,
                        null
                    )
        }
        checkSettingNotNull(calendarStartDayOfWeek)
    }

    private fun setUpReminderNotificationPreferenceValueLoading() {
        loadSettingValue {
            isCheckedReminderNotification =
                userPreferencesRepository.loadReminderNotificationPreference()
                    .map { preference ->
                        preference.isChecked
                    }.stateIn(
                        viewModelScope,
                        SharingStarted.Eagerly,
                        null
                    )
        }
        checkSettingNotNull(isCheckedReminderNotification)

        loadSettingValue {
            reminderNotificationTime =
                userPreferencesRepository.loadReminderNotificationPreference()
                    .map { preference ->
                        preference.notificationLocalTime
                    }.stateIn(
                        viewModelScope,
                        SharingStarted.Eagerly,
                        null
                    )
        }
        checkSettingNotNull(reminderNotificationTime)
    }

    private fun setUpPasscodeLockPreferenceValueLoading() {
        loadSettingValue {
            isCheckedPasscodeLock =
                userPreferencesRepository.loadPasscodeLockPreference()
                    .map { preference ->
                        preference.isChecked
                    }.stateIn(
                        viewModelScope,
                        SharingStarted.Eagerly,
                        null
                    )
        }
        checkSettingNotNull(isCheckedPasscodeLock)

        loadSettingValue {
            passcode =
                userPreferencesRepository.loadPasscodeLockPreference()
                    .map { preference ->
                        preference.passCode
                    }.stateIn(
                        viewModelScope,
                        SharingStarted.Eagerly,
                        null
                    )
        }
        checkSettingNotNull(passcode)
    }

    private fun setUpWeatherInfoAcquisitionPreferenceValueLoading() {
        loadSettingValue {
            isCheckedWeatherInfoAcquisition =
                userPreferencesRepository.loadWeatherInfoAcquisitionPreference()
                    .map { preference ->
                        preference.isChecked
                    }.stateIn(
                        viewModelScope,
                        SharingStarted.Eagerly,
                        null
                    )
        }
        checkSettingNotNull(isCheckedWeatherInfoAcquisition)
    }

    private fun interface SettingLoadingProcess {
        @Throws(Throwable::class)
        fun load()
    }

    private fun loadSettingValue(
        loadingProcess: SettingLoadingProcess
    ) {
        try {
            loadingProcess.load()
        } catch (e: Throwable) {
            Log.e(logTag, "アプリ設定値読込_失敗", e)
            addSettingLoadingErrorMessage()
        }
    }

    private fun <T> checkSettingNotNull(setting: StateFlow<T?>) {
        viewModelScope.launch(Dispatchers.IO) {
            setting.collect { value: T? ->
                if (_isAllSettingsNotNull.value) return@collect
                if (value == null) return@collect

                checkAllSettingsNotNull()
            }
        }
    }

    private fun checkAllSettingsNotNull() {
        _isAllSettingsNotNull.value =
            themeColor.value != null
                    && calendarStartDayOfWeek.value != null
                    && isCheckedReminderNotification.value != null
                    && checkIsCheckedReminderNotificationNotNull()
                    && passcode.value != null
                    && isCheckedWeatherInfoAcquisition.value != null
    }

    private fun checkIsCheckedReminderNotificationNotNull(): Boolean {
        val isCheckedReminderNotification = isCheckedReminderNotification.value ?: return false
        if (isCheckedReminderNotification) {
            return reminderNotificationTime.value != null
        }

        return true
    }

    private fun addSettingLoadingErrorMessage() {
        if (equalLastAppMessage(SettingsAppMessage.SettingLoadingFailure)) return  // 設定更新エラー通知の重複防止

        addAppMessage(SettingsAppMessage.SettingLoadingFailure)
    }

    suspend fun saveThemeColor(value: ThemeColor): Boolean {
        val preferenceValue = ThemeColorPreference(value)
        return updateSettingValue{
            userPreferencesRepository.saveThemeColorPreference(preferenceValue)
        }
    }

    suspend fun saveCalendarStartDayOfWeek(value: DayOfWeek): Boolean {
        val preferenceValue =
            CalendarStartDayOfWeekPreference(value)
        return updateSettingValue{
            userPreferencesRepository.saveCalendarStartDayOfWeekPreference(preferenceValue)
        }
    }

    suspend fun saveReminderNotificationValid(value: LocalTime): Boolean {
        val preferenceValue =
            ReminderNotificationPreference(true, value)
        return updateSettingValue{
            userPreferencesRepository.saveReminderNotificationPreference(preferenceValue)
            workerRepository.registerReminderNotificationWorker(value)
        }
    }

    suspend fun saveReminderNotificationInvalid(): Boolean {
        val preferenceValue =
            ReminderNotificationPreference(false, null as LocalTime?)
        return updateSettingValue{
            userPreferencesRepository.saveReminderNotificationPreference(preferenceValue)
            workerRepository.cancelReminderNotificationWorker()
        }
    }

    suspend fun savePasscodeLock(value: Boolean): Boolean {
        val passcode = if (value) {
            "0000" // TODO:仮
        } else {
            ""
        }

        val preferenceValue = PassCodeLockPreference(value, passcode)
        return updateSettingValue{
            userPreferencesRepository.savePasscodeLockPreference(preferenceValue)
        }
    }

    suspend fun saveWeatherInfoAcquisition(value: Boolean): Boolean {
        val preferenceValue =
            WeatherInfoAcquisitionPreference(value)
        return updateSettingValue{
            userPreferencesRepository.saveWeatherInfoAcquisitionPreference(preferenceValue)
        }
    }

    private fun interface SettingUpdateProcess {
        @Throws(
            IOException::class,
            Exception::class
        )
        suspend fun update()
    }

    private suspend fun updateSettingValue(
        updateProcess: SettingUpdateProcess
    ): Boolean {
        try {
            updateProcess.update()
        } catch (e: IOException) {
            Log.e(logTag, "アプリ設定値更新_失敗", e)
            addSettingUpdateErrorMessage()
            return false
        } catch (e: Exception) {
            Log.e(logTag, "アプリ設定値更新_失敗", e)
            addSettingUpdateErrorMessage()
            return false
        }
        return true
    }

    private fun addSettingUpdateErrorMessage() {
        if (equalLastAppMessage(SettingsAppMessage.SettingUpdateFailure)) return  // 設定更新エラー通知の重複防止

        addAppMessage(SettingsAppMessage.SettingUpdateFailure)
    }

    fun updateGeoCoordinates(geoCoordinates: GeoCoordinates) {
        _geoCoordinates.value = geoCoordinates
    }

    fun clearGeoCoordinates() {
        _geoCoordinates.value = initialGeoCoordinates
    }

    suspend fun deleteAllDiaries(): Boolean {
        try {
            diaryRepository.deleteAllDiaries()
        } catch (e: Exception) {
            Log.e(logTag, "全日記削除_失敗", e)
            addAppMessage(SettingsAppMessage.AllDiaryDeleteFailure)
            return false
        }
        return true
    }

    suspend fun initializeAllSettings(): Boolean {
        return updateSettingValue{
            userPreferencesRepository.initializeAllPreferences()
        }
    }

    suspend fun deleteAllData(): Boolean {
        try {
            diaryRepository.deleteAllData()
        } catch (e: Exception) {
            Log.e(logTag, "アプリ全データ削除_失敗", e)
            addAppMessage(SettingsAppMessage.AllDataDeleteFailure)
            return false
        }
        return initializeAllSettings()
    }
}
